File: arena-jface/src/main/java/org/uqbar/lacar/ui/impl/jface/tree/TreeContentProvider.java
package org.uqbar.lacar.ui.impl.jface.tree;

import java.util.List;

import org.eclipse.core.databinding.observable.set.IObservableSet;
import org.eclipse.core.databinding.observable.value.IObservableValue;
import org.eclipse.core.databinding.observable.value.IValueChangeListener;
import org.eclipse.core.databinding.observable.value.ValueChangeEvent;
import org.eclipse.jface.viewers.ITreeContentProvider;
import org.eclipse.jface.viewers.Viewer;
import org.uqbar.commons.utils.ReflectionUtils;
import org.uqbar.lacar.ui.impl.jface.bindings.JFaceObservableFactory;

public class TreeContentProvider  implements ITreeContentProvider {
	
	private IObservableValue observableParentValue;
	private IObservableSet observableChildrenValue;
	private String parentPropertyName;
	private String childrenPropertyName;
	

	public TreeContentProvider(String parentPropertyName, String childrenPropertyName) {
		this.parentPropertyName = parentPropertyName;
		this.childrenPropertyName = childrenPropertyName;
	}


	public Object[] getChildren(Object parentElement) {
		return ((List<?>) ReflectionUtils.invokeGetter(parentElement, childrenPropertyName)).toArray();
	}

	public Object getParent(Object element) {
		return ReflectionUtils.invokeGetter(element, parentPropertyName);
	}

	public boolean hasChildren(Object element) {
		return true;
	}


	@Override
	public void dispose() {
		if (this.observableParentValue != null) {
			this.observableParentValue.dispose();
		}
		if (this.observableChildrenValue != null) {
			this.observableChildrenValue.dispose();
		}
	}
	@Override
	public void inputChanged(final Viewer viewer, Object oldInput, Object newInput) {
		// Dispose old value observer
				if (this.observableChildrenValue != null) {
					this.observableChildrenValue.dispose();
				}

				// Create new observer and listen to its changes.
				if (newInput != null) {
					this.observableChildrenValue = JFaceObservableFactory.observeSet(newInput, this.childrenPropertyName);
					this.observableParentValue = JFaceObservableFactory.observeProperty(newInput, this.parentPropertyName);
					this.observableParentValue.addValueChangeListener(new IValueChangeListener() {
						@Override
						public void handleValueChange(ValueChangeEvent event) {
							viewer.refresh();
						}
					});
				}
				else {
					this.observableChildrenValue = null;
				}
	}


	@Override
	public Object[] getElements(Object inputElement) {
		return getChildren(inputElement);
	}
}

File: arena-jface/src/main/scala/org/uqbar/lacar/ui/impl/jface/builder/tree/TreeContentProvider.scala
package org.uqbar.lacar.ui.impl.jface.builder.tree

import org.eclipse.jface.viewers.ITreeContentProvider
import org.eclipse.core.databinding.observable.value.IObservableValue
import org.eclipse.core.databinding.observable.set.IObservableSet
import org.uqbar.commons.utils.ReflectionUtils._
import java.util.List
import org.eclipse.jface.viewers.Viewer
import org.uqbar.arena.jface.JFaceImplicits._
import org.eclipse.jface.databinding.swt.SWTObservables._
import org.eclipse.core.databinding.beans.BeansObservables
import org.uqbar.lacar.ui.impl.jface.bindings.JFaceObservableFactory._

/**
 * 
 */
class TreeContentProvider(parentPropertyName:String, childrenPropertyName:String) extends ITreeContentProvider {
	private var observableParentValue : IObservableValue = _
	private var observableChildrenValue : IObservableSet = _

	//TODO: Esto no se banca colecciones de scala
	override def getChildren(parentElement:Object) = { 
		invokeGetter(parentElement, childrenPropertyName).asInstanceOf[List[_]].toArray
	}

	override def getParent(element:Object) = invokeGetter(element, parentPropertyName)
	override def getElements(inputElement:Object) = getChildren(inputElement)
	override def hasChildren(element:Object) = true

	override def dispose() = {
		disposeParentValue
		if (observableChildrenValue != null) observableChildrenValue dispose
	}

	def disposeParentValue() = if (observableParentValue != null) observableParentValue dispose

	override def inputChanged(viewer:Viewer, oldInput:Object, newInput:Object) = {
		// Dispose old value observer
		disposeParentValue

		// Create new observer and listen to its changes.
		if (newInput != null) {
			observableChildrenValue = observeSet(newInput, childrenPropertyName)
			observableParentValue = observeProperty(newInput, parentPropertyName)
			observableParentValue.addValueChangeListener{() => viewer.refresh }
		}
		else {
			this.observableChildrenValue = null
		}
	}

}